### PR TITLE
[Behat] Delete objects related to order during order deletion

### PR DESCRIPTION
**Pull request by lchrusciel**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | #4140
| License       | MIT

Depends on #4235

### CODE REVIEW

File: src/Sylius/Behat/Context/Domain/PaymentContext.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        $this->paymentRepository = $paymentRepository;
+    }
+
+    /**
+     * @Then /^there should be no ("[^"]+" payments) in the registry$/
+     */
+    public function orderShouldNotExistInTheRegistry(PaymentMethodInterface $paymentMethod)
+    {
+        $payment = $this->paymentRepository->findByPaymentMethod($paymentMethod);
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Domain/PaymentContext.php]: $payment**s**

File: src/Sylius/Behat/spec/Context/Domain/PaymentContextSpec.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+        $this->orderShouldNotExistInTheRegistry($cashOnDeliveryPaymentMethod);
+    }
+
+    function it_throws_an_exception_if_payment_still_exist(
+        PaymentRepositoryInterface $paymentRepository,
+        PaymentMethodInterface $paypalPaymentMethod,
+        PaymentInterface $payment
+    ) {
+        $paymentRepository->findByPaymentMethod($paypalPaymentMethod)->willReturn($payment);
```
Comment (lchrusciel) [inline src/Sylius/Behat/spec/Context/Domain/PaymentContextSpec.php]: **[**$payment**]**

File: src/Sylius/Behat/Context/Setup/CustomerContext.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    public function __construct(RepositoryInterface $customerRepository, FactoryInterface $customerFactory)
+    {
+        $this->customerRepository = $customerRepository;
+        $this->customerFactory = $customerFactory;
+    }
+
+    /**
+     * @Transform :customer
+     */
+    public function getCustomerByEmail($customerEmail)
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Setup/CustomerContext.php]: IMO `($email)` is enough.

File: src/Sylius/Bundle/PaymentBundle/Doctrine/ORM/PaymentRepository.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+use Sylius\Component\Payment\Model\PaymentMethodInterface;
+use Sylius\Component\Payment\Repository\PaymentRepositoryInterface;
+
+/**
+ * @author Łukasz Chruściel <lukasz.chrusciel@lakion.com>
+ */
+class PaymentRepository extends EntityRepository implements PaymentRepositoryInterface
+{
+    /**
+     * @param PaymentMethodInterface $paymentMethod
```
Comment (lchrusciel) [inline src/Sylius/Bundle/PaymentBundle/Doctrine/ORM/PaymentRepository.php]: {@inheritdoc}

File: src/Sylius/Component/Payment/Repository/PaymentRepositoryInterface.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+namespace Sylius\Component\Payment\Repository;
```
Comment (lchrusciel) [inline src/Sylius/Component/Payment/Repository/PaymentRepositoryInterface.php]: Missing blank line

File: src/Sylius/Behat/spec/Context/Setup/CustomerContextSpec.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+    function it_creates_new_customer_if_it_does_not_exist(
+        CustomerInterface $customer,
+        FactoryInterface $customerFactory,
+        RepositoryInterface $customerRepository
+    ) {
+        $customerRepository->findOneBy(['email' => 'oliver.queen@star.com'])->willReturn(null);
+
+        $customerFactory->createNew()->willReturn($customer);
+        $customer->setEmail('oliver.queen@star.com');
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/spec/Context/Setup/CustomerContextSpec.php]: Missing should be called

File: src/Sylius/Behat/Context/Domain/PaymentContext.php [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+     */
+    public function __construct(PaymentRepositoryInterface $paymentRepository)
+    {
+        $this->paymentRepository = $paymentRepository;
+    }
+
+    /**
+     * @Then /^there should be no ("[^"]+" payments) in the registry$/
+     */
+    public function orderShouldNotExistInTheRegistry(PaymentMethodInterface $paymentMethod)
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Domain/PaymentContext.php]: **payment**ShouldNotExistInTheRegistry

File: src/Sylius/Behat/spec/Context/Setup/OrderContextSpec.php [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        $order->setCurrency($currency)->shouldBeCalled();
+        $order->setNumber('#00000022')->shouldBeCalled();
+
+        $orderRepository->add($order)->shouldBeCalled();
+        $sharedStorage->set('order', $order)->shouldBeCalled();
+
+        $this->theCustomerPlacedAnOrder($customer, '#00000022');
+    }
+
+    function it_sets_options_chosen_by_customer(
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/spec/Context/Setup/OrderContextSpec.php]: Maybe better `it_adds_shipping_payment_and_addressing_info_to_an_order`?

File: src/Sylius/Behat/spec/Context/Domain/PaymentContextSpec.php [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        $this->shouldHaveType('Sylius\Behat\Context\Domain\PaymentContext');
+    }
+
+    function it_implements_context_interface()
+    {
+        $this->shouldImplement(Context::class);
+    }
+
+    function it_checks_if_an_order_exists_in_repository(
```
Comment (lchrusciel) [inline src/Sylius/Behat/spec/Context/Domain/PaymentContextSpec.php]: it_checks_if_an_**payment**_exists_in_repository

File: src/Sylius/Behat/spec/Context/Setup/OrderContextSpec.php [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+        $payment->setMethod($paymentMethod)->shouldBeCalled();
+        $shipment->setMethod($shippingMethod)->shouldBeCalled();
+        $orderShipmentFactory->createForOrder($order)->shouldBeCalled();
+
+        $objectManager->flush()->shouldBeCalled();
+
+        $this->theCustomerChoseShippingToWithPayment($shippingMethod, $address, $paymentMethod);
+    }
+
+    function it_sets_items_bought_by_customer(
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/spec/Context/Setup/OrderContextSpec.php]: it_**adds_item_**bought...?

File: src/Sylius/Behat/spec/Context/Setup/PaymentContextSpec.php [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    function it_converts_payment_method_name_into_payment_method_object(
+        RepositoryInterface $paymentMethodRepository,
+        PaymentMethod $paymentMethod
+    ) {
+        $paymentMethodRepository->findOneBy(['name' => 'Offline'])->willReturn($paymentMethod);
+
+        $this->getPaymentMethodByName('Offline')->shouldReturn($paymentMethod);
+    }
+
+    function it_throws_element_not_found_exception_if_payment_method_has_been_not_found (
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/spec/Context/Setup/PaymentContextSpec.php]: has_**not_been**_found

File: src/Sylius/Bundle/PaymentBundle/Doctrine/ORM/PaymentRepository.php [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+     *
+     * @return PaymentInterface[]
+     */
+    public function findByPaymentMethod(PaymentMethodInterface $paymentMethod)
+    {
+        return $this
+            ->createQueryBuilder('o')
+            ->addSelect('method')
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/PaymentBundle/Doctrine/ORM/PaymentRepository.php]: Do we need this select?

File: src/Sylius/Behat/Context/Setup/AddressingContext.php [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+/**
+ * @author Łukasz Chruściel <lukasz.chrusciel@lakion.com>
+ */
+class AddressingContext implements Context
+{
+    /**
+     * @var FactoryInterface
+     */
+     private $addressFactory;
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Setup/AddressingContext.php]: Wrong indent.

File: src/Sylius/Behat/Context/Setup/PaymentContext.php [hunk 14]
```diff
@@ -76,4 +78,18 @@ public function storeAllowsPaying($paymentMethodName)
...
+
+    /**
+     * @Transform /^"([^"]+)" payment$/
+     * @Transform /^"([^"]+)" payments$/
+     */
+    public function getPaymentMethodByName($paymentMethodName)
+    {
+        $paymentMethod = $this->paymentMethodRepository->findOneBy(['name' => $paymentMethodName]);
+        if (null === $paymentMethod) {
+            throw new \InvalidArgumentException(sprintf('Payment method with name "%s" does not exist', $paymentMethodName));
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Setup/PaymentContext.php]: Missing dot at the end of exception message :smile:

File: src/Sylius/Behat/Context/Setup/PaymentContext.php [hunk 15]
```diff
@@ -76,4 +78,18 @@ public function storeAllowsPaying($paymentMethodName)
 
         $this->paymentMethodRepository->add($paymentMethod);
     }
+
+    /**
+     * @Transform /^"([^"]+)" payment$/
+     * @Transform /^"([^"]+)" payments$/
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Setup/PaymentContext.php]: What is point of that?

File: src/Sylius/Behat/spec/Context/Setup/PaymentContextSpec.php [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+{
+    function let(
+        RepositoryInterface $paymentMethodRepository,
+        SharedStorageInterface $sharedStorage,
+        FactoryInterface $paymentMethodFactory,
+        PaymentMethodNameToGatewayConverterInterface $paymentMethodNameToGatewayConverter
+    ) {
+        $this->beConstructedWith($paymentMethodRepository, $sharedStorage, $paymentMethodFactory, $paymentMethodNameToGatewayConverter);
+    }
+    function it_is_initializable()
```
Comment (Zales0123) [inline src/Sylius/Behat/spec/Context/Setup/PaymentContextSpec.php]: Missing blank line before.

**Comment by michalmarcinkowski** [comment]
`deletedAt` field should be removed from order and then database migration will also be required

**Comment by lchrusciel** [comment]
@michalmarcinkowski done

File: src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Order.orm.xml [hunk 17]
```diff
@@ -20,17 +20,19 @@
             <join-column name="channel_id" referenced-column-name="id" nullable="true" />
         </many-to-one>
 
-        <many-to-one field="shippingAddress" target-entity="Sylius\Component\Addressing\Model\AddressInterface">
+        <many-to-one field="shippingAddress" target-entity="Sylius\Component\Addressing\Model\AddressInterface" orphan-removal="true">
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Order.orm.xml]: This is unidirectional relation, are we sure that orphan removal works here at all?

File: src/Sylius/Behat/Context/Setup/OrderContext.php [hunk 18]
```diff
@@ -135,6 +143,8 @@ public function theCustomerChoseShippingToWithPayment(
         $this->orderShipmentFactory->processOrderShipment($order);
         $order->getShipments()->first()->setMethod($shippingMethod);
 
+        $this->shippingChargesProcessor->applyShippingCharges($order);
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Setup/OrderContext.php]: Let's remember to use Recalculator after @NeverResponse's PR is merged.

**Comment by pjedrzejewski** [comment]
Thanks Łukasz!

**Pull request merged**